packages/gatsby-theme-docs/src/layouts/internals/layout-application.js

```diff
@@ -41,18 +43,55 @@ const Container = styled.div`
   }
 `;
 
-const LayoutApplication = (props) => (
-  <>
-    <Root
-      role="application"
-      id="application"
-      isGlobalNotificationVisible={Boolean(props.globalNotification)}
-    >
-      <Container {...props} />
-    </Root>
-    <div id="modal-portal" />
-  </>
-);
+const localStorageProvider = () => {
+  // When initializing, we restore the data from `localStorage` into a map.
+  const map = new Map(JSON.parse(localStorage.getItem('app-cache') || '[]'));
+
+  const handleBeforUnload = (event) => {
+    const appCache = JSON.stringify(Array.from(map.entries()));
+    localStorage.setItem('app-cache', appCache);
+  };
+
+  const handleBeforeAuth0LoginRedirect = (event) => {
+    // I'm about to redirect to auth0, I want to disable the set cache logic
+    window.removeEventListener(handleBeforUnload);
+    // and ensure the cache is clean
+    localStorage.removeItem('app-cache');
+  };
+
+  // Before unloading the app, we write back all the data into `localStorage`.
+  window.addEventListener('beforeunload', handleBeforUnload);
+
+  window.addEventListener(
+    'beforeAuth0LoginRedirect',
+    handleBeforeAuth0LoginRedirect
+  );
+
+  // We still use the map for write & read for performance.
+  return map;
+};
+
+const LayoutApplication = (props) => {
+  const { isClientSide } = useIsClientSide();
+  return (
+    <>
+      <Root
+        role="application"
+        id="application"
+        isGlobalNotificationVisible={Boolean(props.globalNotification)}
+      >
+        {isClientSide ? (
+          <SWRConfig value={{ provider: localStorageProvider }}>
```

Comment:
First doubt. If we want to have the course/topic status available in the sidebar and (potentially in the future) in the actual content page, the only place common between these 2 layouts is the application layout.
Adding cache provider here means that all the usages of swr in the app will be cached in local cache. This might be what we actually want, or might be creating side effects.

Reply:
#1  I wonder if something like the SWRConfig would not be better placed in gatsby-browser's "wrapXXXElement" hooks.   That implicitly makes them client side only and also makes sure they are in all layouts. 

#2  having one cache provider for any SWR even across microsites felt weird to me too in the beginning but I concluded that it's the whole point of it.   The cache keys are the URLs you use to fetch from SWR and the content is the API response.  And we want that information to be available at least in stale state for other pages / views / microsites.   So it must be one fixed key like "app-cache" and also at top level for the domain.  We could use a more specific localstorage key maybe that is more explanatory.